### ISSUE TITLE
Platform usage stats for canarie

### ISSUE THREAD

**Issue by huard** [Ouranosinc/pavics-sdi#118]
https://pavics.ouranos.ca/canarie/pavics/platform/stats

Says 0 usage.

**Comment by dbyrns** [comment]
Here is some clues that could help you on that issue:
A cron job (https://github.com/Ouranosinc/CanarieAPI/blob/master/canarieapi-cron) is parsing the nginx log (https://github.com/Ouranosinc/PAVICS/blob/master/birdhouse/config/proxy/nginx.conf#L22)
Note that the default access.log file is not used as it is redirect to stdout using symlink by the base image, it's important to keep the access_file.log which is parsed (and rotated to avoid counting twice the same access) by the cron job.
This is the regex to be matched in the log file to increment the access stats (store in a db) :  
https://github.com/Ouranosinc/PAVICS/blob/master/birdhouse/config/canarie-api/docker_configuration.py.template#L26
Right now we're looking at magpie, geoserver and twitcher (which should proxy every wms, wps access).
My first step would be to look at boreas nginx access logs and make sure that access are effectively logged and that some of the access fit the current regex.
Also, since /jupiter and /raven routes have been added and probably accessed directly, you may want to add them in the regex so that they are taken into account in the stats.

**Comment by huard** [comment]
Merci, @tlvu could you look into this ?

**Comment by tlvu** [comment]
Still need /raven (no access control) for demo or I can remove it?  This is higher prio than the raven doc that I already started?

**Comment by huard** [comment]
Nope, thanks !
No, this is medium priority.

**Comment by huard** [comment]
We need this to work by mid-September.

**Comment by tlvu** [comment]
> My first step would be to look at boreas nginx access logs and make sure that access are effectively logged and that some of the access fit the current regex.

```
$ ssh mourad@pavics.ouranos.ca docker exec proxy ls -lart /var/log/nginx
total 244760
lrwxrwxrwx. 1 root root        11 Nov  4  2017 error.log -> /dev/stderr
lrwxrwxrwx. 1 root root        11 Nov  4  2017 access.log -> /dev/stdout
drwxr-xr-x. 1 root root        27 Jan 12  2018 ..
drwxr-xr-x. 1 root root        37 Jul 12 20:22 .
-rw-r--r--. 1 root root 250629779 Aug  7 19:37 access_file.log
```

**Comment by tlvu** [comment]
I found out why there are no stats: the cron jobs are not running !
```
$ ssh mourad@pavics.ouranos.ca docker exec proxy cat /etc/cron.d/canarieapi-cron
* * * * * root python -c 'from canarieapi import logparser; logparser.cron_job()' >> /var/log/logparser_cron.log 2>&1
* * * * * root python -c 'from canarieapi import monitoring; monitoring.cron_job()' >> /var/log/monitoring_cron.log 2>&1

$ ssh mourad@pavics.ouranos.ca docker exec proxy ls -la /var/log |grep _cron.log
# no logs file found, meaning the cron jobs did not run
```

**Comment by tlvu** [comment]
Deleting and recreating the proxy container on my Vagrant VM seems to get the cron job going again:
```
$ vagrant ssh -- docker exec proxy tail /var/log/logparser_cron.log                                                                         
[2019-08-07 20:50:01,848] [26] [INFO] canarieapi.app_object : Rotating /var/log/nginx/access_file.log
[2019-08-07 20:50:01,851] [26] [INFO] canarieapi.app_object : Asking nginx to reload log file (using pid file : /var/run/nginx.pid
[2019-08-07 20:50:01,851] [26] [INFO] canarieapi.app_object : nginx pid is 10
[2019-08-07 20:50:01,851] [26] [INFO] canarieapi.app_object : Sending USR1 (to reload log) signal to nginx process
[2019-08-07 20:50:02,858] [26] [INFO] canarieapi.app_object : Loading configuration
[2019-08-07 20:50:02,860] [26] [INFO] canarieapi.app_object : Loading log file : /var/log/nginx/access_file.log.1
[2019-08-07 20:50:02,860] [26] [INFO] canarieapi.app_object : Compiling stats from 0 records
[2019-08-07 20:50:02,860] [26] [INFO] canarieapi.app_object : Updating databse
[2019-08-07 20:50:02,860] [26] [DEBUG] canarieapi.app_object : Using db filename : /opt/local/src/CanarieAPI/stats.db
[2019-08-07 20:50:03,032] [26] [INFO] canarieapi.app_object : Done
```

**Comment by tlvu** [comment]
Since I do have what looks like normal stats on my Vagrant VM (same docker image, same config), I think this is an environment issue on Boreas and Medus that the cron job inside the docker container can not start.

Not the first time Boreas and Medus behave weirdly.  The last time was https://github.com/Ouranosinc/PAVICS/issues/138 where I was not able to "wget any httpS URL" inside docker container running on Boreas and Medus.

Differences between my Vagrant VM and Boreas/Medus:
* My Vagrant VM is Ubuntu Bionic 18.04 while Boreas/Medus is Centos 7.3
* My Vagrant VM has a more recent kernel 4.15 while Boreas/Medus is 3.10 due to old Centos 7
* My Vagrant VM has latest docker and docker-compose versions
* My Vagrant VM do not have SELinux enabled

At this stage I think @moulab88 need to take a look at the environment on Boreas/Medus.

**Comment by tlvu** [comment]
@huard @dbyrns By the way, this stat usage count also include the hits from the monitoring itself !

My server is running idle without any notebooks hitting it and the stats count keeps incrementing.  Is this how we have defined this stat?

Also the DB file where the stats are saved are not persisted outside of the container.  If the container is deleted (because of new image deployment or because we need to do it), all the stats are gone.  Is this also intended?

**Comment by tlvu** [comment]
FYI, needed this PR https://github.com/Ouranosinc/PAVICS/pull/153 to get stats working on my Vagrant VM with self-signed SSL cert and pagekite.

**Comment by huard** [comment]
Status ?

**Comment by tlvu** [comment]
Discussed with @moulab88, he'll take a look next week.  Suggested him to start with SELinux and firewall rules.  To have a more recent kernels/docker/docker-compose, might require a full system upgrade ... much more costly.

**Comment by moulab88** [comment]
![image](https://user-images.githubusercontent.com/25254840/63379490-6b8b1880-c362-11e9-8bb0-f56bd76599c5.png)
![image](https://user-images.githubusercontent.com/25254840/63379593-a4c38880-c362-11e9-9a20-5d8cf0bc2dd0.png)

**Comment by huard** [comment]
We have the same issue with the services (raven, indices, etc), where stats are set to 0. 
https://pavics.ouranos.ca/canarie/raven/service/stats

**Comment by tlvu** [comment]
Super @moulab88!  Care to explain what you did to make it work?

**Comment by moulab88** [comment]
Just a new edition of /etc/cron.d/canarie api-cron with vi. It has "unblocked" the cron, maybe too much spaces.
![image](https://user-images.githubusercontent.com/25254840/63444338-667ea580-c404-11e9-90ee-f956f3049f5d.png)

**Comment by huard** [comment]
Should I close this ticket and open a new one for services stats ? Or it's the same underlying issue ?

**Comment by tlvu** [comment]
> Just a new edition of /etc/cron.d/canarie api-cron with vi. It has "unblocked" the cron, maybe too much spaces.

If I am not mistaken, that cron file is part of the image itself so as soon as you'll delete the container and start a new one, your change is gone.

Can you commit a change to the PAVICS repo (put that modified file in a config folder, volume mount it into the container to override the one from the image) to make your change permanent?

**Comment by tlvu** [comment]
> Should I close this ticket and open a new one for services stats ? Or it's the same underlying issue ?

Had a debugging session with @moulab88 today.  It's not the same underlying issue (cron job not running).

It's a different root cause: each route regex should not match other route regexes because as soon as a log line match one route regex, it skips all the remaining route regexes !

We had the "node" route `'/(magpie|twitcher|geoserver)/.*'` which basically eat all the other routes under `/twitcher/ows/proxy/...`.

If we need to have a generic catch all route, it has to be the last element.  The "node" route above was the first element !

**Comment by huard** [comment]
@tlvu Could you then close this ticket and open another one ? This is something we could assign to CRIM if you think it would help.

**Comment by tlvu** [comment]
@huard waiting for a commit from @moulab88 to close this issue.

**Comment by tlvu** [comment]
PR https://github.com/Ouranosinc/PAVICS/pull/150 "Upgrade thredds and use upstream image from Unidata" will also close this issue.

**Comment by tlvu** [comment]
> If we need to have a generic catch all route, it has to be the last element. The "node" route above was the first element !

Been thinking about this, what I said above is wrong.  Python dict is not ordered so there's no guaranty that the last route will be match last so we can not have a "generic catch all" route at the end.

We should ensure all route regexes are all distinct from each other.  Each log line should match 1 single route regex, no more.

**Comment by tlvu** [comment]
> Could you then close this ticket and open another one ? This is something we could assign to CRIM if you think it would help.

Yes I think a different ticket to review all the route regexes to make them all distinct is in order.

Feel free to assign to the person that knows which service should match which route precisely.  I only knows the list of routes from our proxy definition.  If there are more specialized routes, with special wps query params, I might not know.
